ql/src/java/org/apache/hadoop/hive/ql/io/orc/VectorizedOrcAcidRowBatchReader.java

```diff
@@ -1605,6 +1618,46 @@ public int compareTo(CompressedOwid other) {
         throw e; // rethrow the exception so that the caller can handle.
       }
     }
+
+    /**
+     * Create delete delta reader. Caching orc tail to avoid FS lookup/reads for repeated scans.
+     *
+     * @param deleteDeltaFile
+     * @param conf
+     * @param fs FileSystem
+     * @return delete file reader
+     * @throws IOException
+     */
+    private Reader getDeleteDeltaReader(Path deleteDeltaFile, JobConf conf, FileSystem fs) throws IOException {
+      OrcTail deleteDeltaTail = deleteDeltaOrcTailCache.getIfPresent(deleteDeltaFile);
```

Comment:
Is the OrcTail thread safe? If I understandcorrectly the tail will be read by multiple LLAP threads concurrenly

Reply:
Yes, it will not change for the file.